presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java

```diff
@@ -277,42 +284,115 @@ public static LongWriteFunction timeWriteFunction()
      * forward offset change (a 'gap'). This includes regular DST changes (e.g. Europe/Warsaw) and one-time policy changes
-     * (Asia/Kathmandu's shift by 15 minutes on January 1, 1986, 00:00:00).
+     * (Asia/Kathmandu's shift by 15 minutes on January 1, 1986, 00:00:00). If only driver supports {@link LocalDateTime}, use
+     * {@link #timestampWriteFunction} instead.
      */
-    public static LongWriteFunction timestampWriteFunctionUsingSqlTimestamp()
+    @Deprecated
+    public static LongWriteFunction timestampWriteFunctionUsingSqlTimestamp(ConnectorSession connectorSession)
     {
+        if (connectorSession.isLegacyTimestamp()) {
+            ZoneId sessionZone = ZoneId.of(connectorSession.getTimeZoneKey().getId());
+            return (statement, index, value) -> statement.setTimestamp(index, Timestamp.valueOf(fromPrestoLegacyTimestamp(value, sessionZone)));
+        }
+        return (statement, index, value) -> statement.setTimestamp(index, Timestamp.valueOf(fromPrestoTimestamp(value)));
+    }
+
+    public static LongWriteFunction timestampWriteFunction(ConnectorSession session)
+    {
+        if (session.isLegacyTimestamp()) {
+            ZoneId sessionZone = ZoneId.of(session.getTimeZoneKey().getId());
+            return (statement, index, value) -> statement.setObject(index, fromPrestoLegacyTimestamp(value, sessionZone));
+        }
         return (statement, index, value) -> {
-            // Copied from `QueryBuilder.buildSql`
-            // TODO verify correctness, add tests and support non-legacy timestamp
-            statement.setTimestamp(index, new Timestamp(value));
+            statement.setObject(index, fromPrestoTimestamp(value));
         };
     }
 
-    public static Optional<ColumnMapping> jdbcTypeToPrestoType(JdbcTypeHandle type)
+    /**
+     * @deprecated applicable in legacy timestamp semantics only
+     */
+    @Deprecated
+    private static long toPrestoLegacyTimestamp(LocalDateTime localDateTime, ZoneId sessionZone)
+    {
+        return localDateTime.atZone(sessionZone)
+                .toInstant()
+                .toEpochMilli();
+    }
+
+    private static long toPrestoTimestamp(LocalDateTime localDateTime)
+    {
+        // Equivalent to `localDateTime.atZone(UTC).toInstant().toEpochMilli()`
+        return SECONDS.toMillis(localDateTime.toEpochSecond(ZoneOffset.UTC)) + NANOSECONDS.toMillis(localDateTime.getNano());
+    }
+
+    /**
+     * @deprecated applicable in legacy timestamp semantics only
+     */
+    @Deprecated
+    private static LocalDateTime fromPrestoLegacyTimestamp(long value, ZoneId sessionZone)
+    {
+        return Instant.ofEpochMilli(value)
+                .atZone(sessionZone)
+                .toLocalDateTime();
+    }
+
+    private static LocalDateTime fromPrestoTimestamp(long value)
+    {
+        // Equivalent to `Instant.ofEpochMilli(value).atZone(ZoneOffset.UTC).toLocalDateTime()`
```

Comment:
This seems complicated and error prone. Why not use the simpler version?

Reply:
Do you mean to replace `.... MILLISECONDS.toSeconds(value < 0 ? value - 1000 : value),
                toIntExact(MILLISECONDS.toNanos(floorMod(value, 1000))) ...` math with `Instant.ofEpochMilli(value).atZone(ZoneOffset.UTC).toLocalDateTime()` ?

I feel pretty confident about correctness (only) thanks to the code coverage, but I can still change to use the simpler version.

Reply:
Yes, this fits with our philosophy of writing simpler code unless benchmarks demonstrate a need for something more complicated. (my guess is that this is meaningless given all the overhead of JDBC)